modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,369 @@
+		numMerkleRoots int
+	}
+
+	// cachedSubTree is a cached subTree of a merkle tree. A height of 0 means
+	// that the sum is the hash of a leaf. A subTree of height 1 means sum is
+	// the root of 2 leaves. A subTree of height 2 contains the root of 4
+	// leaves and so on.
+	cachedSubTree struct {
+		height int         // height of the subTree
+		sum    crypto.Hash // root of the subTree
+	}
+)
+
+// loadExistingMerkleRoots reads creates a merkleRoots object from existing
+// merkle roots.
+func loadExistingMerkleRoots(file *fileSection) (*merkleRoots, error) {
+	mr := &merkleRoots{
+		rootsFile: file,
+	}
+	// Get the number of roots stored in the file.
+	var err error
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return nil, err
+	}
+	// Get the size of the file and read it.
+	var size int64
+	if size, err = file.Size(); err != nil {
+		return nil, err
+	}
+	fileData := make([]byte, size)
+	if _, err := file.ReadAt(fileData, 0); err != nil {
+		return nil, err
+	}
+	// Read the roots from the file without reading all of them at once.
+	r := bytes.NewBuffer(fileData)
+	for i := 0; i < mr.numMerkleRoots; i++ {
+		var root crypto.Hash
+		if _, err = io.ReadFull(r, root[:]); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+
+		// Append the root to the uncachedRoots
+		mr.appendRootMemory(root)
+	}
+	return mr, nil
+}
+
+// newCachedSubTree creates a cachedSubTree from exactly
+// 2^merkleRootsCacheHeight roots.
+func newCachedSubTree(roots []crypto.Hash) *cachedSubTree {
+	// Sanity check the input length.
+	if len(roots) != merkleRootsPerCache {
+		build.Critical("can't create a cached subTree from the provided number of roots")
+	}
+	return &cachedSubTree{
+		height: int(merkleRootsCacheHeight + sectorHeight),
+		sum:    cachedMerkleRoot(roots),
+	}
+}
+
+// newMerkleRoots creates a new merkleRoots object. This doesn't load existing
+// roots from file and will assume that the file doesn't contain any roots.
+// Don't use this on a file that contains roots.
+func newMerkleRoots(file *fileSection) *merkleRoots {
+	return &merkleRoots{
+		rootsFile: file,
+	}
+}
+
+// fileOffsetFromRootIndex calculates the offset of the merkle root at index i from
+// the beginning of the contract file.
+func fileOffsetFromRootIndex(i int) int64 {
+	return crypto.HashSize * int64(i)
+}
+
+// appendRootMemory appends a root to the in-memory structure of the merkleRoots. If
+// the length of the uncachedRoots grows too large they will be compressed into
+// a cachedSubTree.
+func (mr *merkleRoots) appendRootMemory(root crypto.Hash) {
+	mr.uncachedRoots = append(mr.uncachedRoots, root)
+	if len(mr.uncachedRoots) == merkleRootsPerCache {
+		mr.cachedSubTrees = append(mr.cachedSubTrees, newCachedSubTree(mr.uncachedRoots))
+		mr.uncachedRoots = mr.uncachedRoots[:0]
+	}
+}
+
+// delete deletes the sector root at a certain index.
+func (mr *merkleRoots) delete(i int) error {
```

Comment:
This operation isn't WAL-safe, it's not idempotent. If the WAL is going to be calling `delete` from a transaction, it needs to be able to call delete multiple times without causing harm.

the issue is that if we call `delete(5)` twice in a row, we delete two different roots, instead of deleting the same root. If this is to be called from the WAL, the instructions actually have to take 2 inputs:

1. some data to write at a specific index (for the swap) (might need a bool too indicating whether a swap is required)

2. some data indicating which index should be truncated to.

Those values combined will let you perform idempotent deletes.